### PR TITLE
Fixed crashing when changeing setpoint widget if state contains dimensions

### PR DESCRIPTION
**Pull request by cd-tronic**
Signed-off-by: cd-tronic <stuff@cd-tronic.de>

### CODE REVIEW

File: OpenHAB.Windows/Controls/SetpointWidget.xaml.cs [hunk 1]
```diff
@@ -20,7 +21,7 @@ public SetpointWidget()
 
         private void ButtonUp_Click(object sender, RoutedEventArgs e)
         {
-            double value = double.Parse(Widget.Item.State);
+            double value = getDoubleValue(Widget.Item.State);
```
Comment (hoffe86) [inline OpenHAB.Windows/Controls/SetpointWidget.xaml.cs]: @cd-tronic Please rename the function to start in upper case (C# Naming Guidlines).  
getDoubleValue -> GetDoubleValue

File: OpenHAB.Windows/Controls/SetpointWidget.xaml.cs [hunk 2]
```diff
@@ -46,6 +47,19 @@ private void ButtonDown_Click(object sender, RoutedEventArgs e)
             Messenger.Default.Send(new TriggerCommandMessage(Widget.Item, value.ToString(CultureInfo.InvariantCulture)));
         }
 
+        private double getDoubleValue(string state)
+        {
+            string newstate = Regex.Replace(state, "[^0-9,.]", string.Empty);
+            double value = 0;
+            try
+            {
+                value = double.Parse(newstate);
```
Comment (hoffe86) [inline OpenHAB.Windows/Controls/SetpointWidget.xaml.cs]: Use double.TryParse(newState) this returns a boolean which indicates if the value was parsable.

**Review by hoffe86 — CHANGES_REQUESTED** [review]
@cd-tronic Most parts are fine, but please fix this two minor comments

**Comment by cd-tronic** [comment]
Hi @hoffe86 ,

i thought about that problem once again and found out that this might not only be a problem for the setpoint item. In fact every item might be affected by a crash if a unit is added to the value. That's why i think it is more forward approach to implement this in a more unspecific way.

What i could imagine is:
1. Create a 'state' class (not shure how to call it exactly) that holds the value and the unit.
2. Create a converter that converts state to 'state'-class.

Now i see 2 options:

1st option: Create this new class while initializing the WidgetBaseClass (using converter) and hold this inside widget class.

2nd option: Create the state class every time a value-change method is called (i think of setpoint and slider widget. Don't know wich item-types might be also affected by the problem).

I would prefer the 2nd option. What do you think?

Why we need to save the unit on it's own:
When changeing the value the widget is updated with the new value from code-behind. The unit won't be shown at this moment until the widget get's updated from the server again.

What do you think of this implementation approach?

**Comment by hoffe86** [comment]
Hey @cd-tronic 
Yeah seems its a general problem and your idea is good to keep the state and unit separate. 
From my point of view I would already separate the state and unit already in the [OpenHabItem](https://github.com/openhab/openhab-windows/blob/master/Openhab.Core/Model/OpenHABItem.cs )class, which is used to parse the xml response from the server.

There we should extend the ParseNode methode to parse state and unit to separate properties. In the next step we can decide if its required to extend die WidgetBase class with a new property for the state.

Advantage of the approach every time when a updated is received the state is parsed correctly and populated to the UI.

**Comment by cd-tronic** [comment]
Hi @hoffe86 ,

I will have a look at it!

**Comment by hoffe86** [comment]
Hey @cd-tronic,
How are you doing? Any updates?

**Comment by cd-tronic** [comment]
Hi @hoffe86,

I'm so sorry. I didn't had time to do anything as i got a pretty big project in.
It seems like next week should be a bit more relaxed.

**Comment by cd-tronic** [comment]
Hi @hoffe86 ,

hm. i tried to figure out how all these elements work together. But I'm still having a hard time with it.
The ParseNode Function inside OpenHABItem class doesn't seem to do anything. I can't get any breakpoint to trigger inside. In fact it doesn't change anything if i completely comment it.

I tried to find a point where all the items get created and thought it is in the ParseItem function inside the OpenHABWidget class. But even commenting it here doesn't change anything.

Do you have any idea where the items get created?

**Comment by cd-tronic** [comment]
@hoffe86 
any idea?

**Comment by hoffe86** [comment]
@cd-tronic Sorry was a bit busy in last time.
I will have look, which part is responsible for the parsing

**Comment by cd-tronic** [comment]
@hoffe86 i saw you active yesterday. Did you get a chance to look in it?
I will have a look at it again next week.

**Comment by hoffe86** [comment]
@cd-tronic Yes I had a look a few moments ago. 
I found out that the state will be set in the "HandleUpdateItemMessage" (Line 63) in OpenHabItem class, which is responsible to retrieve the latest state update via messaging. 

```
private void HandleUpdateItemMessage(UpdateItemMessage message)
{

    if (message.ItemName != Name)
    {
        return;
    }

    DispatcherHelper.ExecuteOnUIThreadAsync(() =>
    {
      ->  State = message.Value; <-
    });
}
```

The message is triggered in the OpenHabClient class in the "StartItemUpdates" methode at line 304, but I think it make sense to do the parsing within the OpenHabItem class. Instead of assign the new value directly to "State" I would introduce a new methode to parse value into State and a new property Unit.

Does this make sense?

**Comment by cd-tronic** [comment]
Hi @hoffe86 ,

can you please review this PR.
I fixed the issue for now regarding the setpoint widget. But please don't merge it as i want to add the changes to other widgets also. But no more time for today.

Regards,

**Comment by cd-tronic** [comment]
It works for SetPoint and Sliderwidget now. 
I also fixed an issue i came across that sliderwidgets don't use Min and Max values.

Would have been better to create a new PR for this. But in fact i don't know how. Use a new Branch for this?

**Comment by hoffe86** [comment]
@cd-tronic I will have a look on our changes this weekend
From my side it is fine to fix the Min/Max issue in the same branch

**Comment by cd-tronic** [comment]
Hi @hoffe86,

sorry i messed everything up when fixing a Commit where i forgot to add the sign-off.

I'm trying to revert everything to the state of yesterday evening. But i need to figure out how.

I get back when i found out how.

Sorry for now.

**Comment by cd-tronic** [comment]
Hi @hoffe86 

sorry i struggled with adding a Sing-Off to a commit. But looks like everything is fine now.

Regards,

**Comment by hoffe86** [comment]
hey @cd-tronic 
Seems go so far. Please change the branch to release/beta, because I want to publish the change in the beta first to test it with a larger audience.

Afterthat we will integrate the changes in master to publish the changes in the general app.

**Comment by hoffe86** [comment]
hey @cd-tronic 
I changed the branch already to release/beta and so you need just to resolve the merge conflicts.

**Comment by cd-tronic** [comment]
Hi @hoffe86

I will check tomorrow!

Thanks

**Comment by cd-tronic** [comment]
@hoffe86 
please check now :)

**Pull request merged**